AGENT-1250: Simplified cluster operations for disconnected environments (#1821)
openshift/enhancements: enhancements/agent-installer/simplified-registry-less-cluster-operations.md

hosting an external registry made it impractical.

### User Stories

* As an OpenShift administrator, I want to install a cluster into a
  disconnected environment without setting up an external registry (and
  optionally a selected set of OLM operators)
* As an OpenShift administrator, I want to upgrade a cluster into a
  disconnected environment without setting up an external registry (and
  optionally a selected set of OLM operators)
* As an OpenShift administrator, I want to add a node to an existing cluster in
  a disconnected environment without setting up an external registry

### Goals

* Ensure successful installation/upgrade/adding new nodes in disconnected
  environments with minimal requirements and no pre-existing registry
* Offer a simplified installation experience for disconnected environments by
  leveraging existing tools like the Agent-Based Installer, Assisted Installer
  and OpenShift Appliance

### Non-Goals

* Allow the user to install/upgrade any OLM operators

## Proposal

From a very high-level perspective, this proposal consists of building and
releasing an extended RHCOS live ISO that includes a full OCP release payload

> (with a selected subset of OLM operators) and all the necessary

Reviewer: Wanted to clarify my understanding, is the scope of this work to include a selected subset of OLM operators in the release payload but for now not allow them to be installed or upgraded? Is there an eventual goal to integrate this with OLM, eg an expectation that OLM would automatically pick up this registry on start?

A new registry service will be injected via MCO to support providing the
release content stored on disk after the reboot (since the previous ephemeral
environment will not be available anymore).

To allow pulling images from the cluster, an ImageDigestMirrorSet
(IDMS) policy will be added to the cluster manifests - using the `api-int` 
DNS entry - so that the set of registries running within
the control plane nodes will become the nominal pullspec of the current release
payload itself.

Finally, a new `InternalReleaseImage` custom resource, managed by MCO, will
also be added as a cluster manifest to keep track and manage the currently
stored internal release image.

### Upgrade

Upgrading an existing lower version cluster will be performed via the same
extended ISO used for managing the installation. As soon as the user will
attach the extended ISO to one of the control planes a temporary registry
will be created to serve its content.
At this point the user could edit the existing IRI resource to add the new
release to trigger the copy process (from the temporary registry
to each local control plane node registry), managed by the Machine Config

> Daemon (MCD). Once the copy has been completed on all the control plane nodes,

Reviewer 2: It might be helpful to have a singular controller for orchestrating the copy process across the MCDs running on the control plane nodes. This makes it easier to do aggregate several node states into overall conditions and to report error modes at the operator level.

Reviewer 3: This design was inspired by [PinnedImageSet](https://github.com/openshift/enhancements/blob/master/enhancements/machine-config/pin-and-pre-load-images.md) in the hope that much of that can be... if not _reused_, at least used as a template.
IIUC PinnedImageSet does have its own controller running inside the MCO. I don't think the language here was intended to exclude that, so it could probably be clarified slightly.

Reviewer 2: Ah, I see. Agreed that we should clarify if PIS is going to be reused here. 

They're specified at the [MachineConfigPool](https://github.com/openshift/api/blob/master/machineconfiguration/v1/types.go#L454) level and the node level images are represented in the status of the [MachineConfigNode](https://github.com/openshift/api/blob/master/machineconfiguration/v1/types_machineconfignode.go#L170) objects. We should be able to leverage that.  We'd need some sort of translation between InternalReleaseImage <=> PinnedImageSets; maybe at the control-plane level? 

I'm also not sure if PIS currently works in bootstrap/firstboot mode, so that might be something I can try getting clarification from the team on(cc @yuqi-zhang if you have context here). The flow above hints that IRI manifest is added at the end of the installation, so perhaps the MCO translation piece is not relevant at bootstrapping time?

Reviewer 4: Hmm, I guess I'm not too clear on this either. It reads to me that the install flow is completely handled by the agent installer (i.e. it will "set up" the registry and image content from the install iso on all the control plane nodes), and upgrade wise, that responsibility falls onto the MCD via specifications in IRI? Does IRI come into the install process at all through the worker nodes?

Re: upgrade, my read of the process outlined below is that:
1. the user plugs in a usb disk ("attach the extended ISO") to the first control plane node
2. the user writes IRI for that specific node, so that the MCD copies it
3. the user unplugs it, plugs it into the second control plane node
4. write IRI to the second node
repeat until done for control plane, then trigger the upgrade for the whole cluster, where the 3 control plane node reads off itself and then have the workers read off of the control planes?

If that's the case, I think regardless we'd need some daemon process to handle the actual "copy". We probably also need some controller handling the central state though,  to ensure the user has successfully completed all copies and write some status before they can upgrade.

Reviewer 3: > It would certainly be advantageous if we could rely on the IRI and MCO even at install time to create the local registry

On reflection, this wouldn't work for single-node :slightly_frowning_face:

the user could start the usual offline upgrade process via the `oc adm upgrade

> --allow-explicit-upgrade --to-image=<release-image>`.

Reviewer 5: What is the suggested approach to extract the relevant 'release-image' from the ISO?
Or maybe since the ISO is constructed for a specific release image anyway, we can keep it transparent for users? E.g. apply a MachineConfig that starts the upgrade (and detects the right version internally)

Reviewer 6: I've lost track of the specific validations the CVO does but I believe the way it's intended to work today is if you provide a symbolic version it has to be on the graph. As soon as you provide a specific pull spec (--allow-explicit-upgrade) the CVO assumes you know what you're doing, not that that's ever been documented explicitly. There's a lot of blurred lines between CLI side guards and what the API accepts.

@wking can you check me here? what protections are in place here and what do you think would be best way to keep them from making a mistake by potentially selecting the wrong mirrored image when expecting to upgrade? Maybe it's time to move all the client side protections into the ClusterVersion CR and CVO via a new interface?

Reviewer 3: It would be great if we had symbolic versions. What I think we'd need to do that is a mapping of tags to sha digests that is cryptographically signed. If that existed it would be fairly easy to build into the ISO, but the infrastructure to set it up if it doesn't would be potentially daunting. Do you know of any existing artifacts that could act as building blocks here? A brief reading of sigstore docs suggests that it won't help.

Reviewer 7: FWIW: Before OSUS (and oc-mirror) were around, when performing OCP4 upgrades on disconnected environments, they had to be done with `--allow-explicit-upgrade` and that did have the risks you are pointing out here of not properly obeying the graph. In that era, the message used to be that you had to be careful to follow the graph on your own or you risked to not be properly supported.

However, I'd prefer to have safeguard mechanisms like the ones you are considering.

Reviewer 3: Arguably there's a high probability that most disconnected upgrades still follow this path and don't use OSUS, and that we should therefore punt this to a future enhancement rather than block on it now. (I agree that I'd _prefer_ to have it, all things being equal.)

At the end of the upgrade the user may decide to delete the previous release

> entry from the IRI resource. In such case the MCD on each control plane node

Reviewer 5: What is the API for removing a previous release?
The suggestion is that MCD will monitor IRI and handle the removal?
I.e. IRI stores the list of images, and MCD removes every image?
If so, how will we provide the full list of images? Include the oc-mirror mapping in ISO?

will take care of removing the older release payload from the local registry.

### Adding a new node

A node could be added to a target existing cluster using the `oc adm
node-image` [command](https://github.com/openshift/enhancements/blob/master/enhancements/oc/day2-add-nodes.md)
(currently, limited to workers). Even though the IDMS resource guarantees that
the `oc adm node-image create` could successfully generate the node ISO, it
won't be sufficient to support the joining process. For that, it will be
necessary to run the registry on a host network port accessible from outside
the cluster. In addition to that, the registry will be secured via a

> pull-secret. Users will be required to add this port to the api-int load

Reviewer 5: The port should be configurable?
Or should we just guide users the open some constant port (e.g. 5000)?

this port to be accessible from all the nodes, as well as the registry port.

### Workflow Description

The various workflows are briefly summarized below:

#### Installation

1. The user downloads the extended RHCOS ISO from the Red Hat Customer Portal.
2. The user moves the extended RHCOS ISO into the disconnected environment.
3. The user boots all the nodes that will be part of the cluster using the
   downloaded ISO.
   (note: even though not strictly relevant for the current proposal,
   we'll assume that the deployment configuration details are already
   available in the installation environment - either provided statically or
   interactively by the user - such as the rendezvous node selection and cluster
   specifications).
4. The installation process starts automatically and runs unattended until
   completed.

#### Upgrade

1. The user downloads the extended RHCOS ISO from the Red Hat Customer Portal,
   by selecting a compatible newer version release for the target cluster.
2. The user moves the extended RHCOS ISO into the disconnected environment.
3. The user attaches the extended RHCOS ISO to one of the control plane nodes

>    of the target cluster.

Reviewer 7: On one hand, how/where the image is exposed might vary depending on the environment (and the details of the driver used for the virtual or real drive involved).

On the other hand, it would be ideal to abstract the user as much as possible of this kind of host-level details.

So ideally this should be detected by just attaching it (without requiring the user to mount anything manually), yet it might be wise to keep some way for manual override (in case the customer has the cluster in an environment where the ISO is attached in a way that cannot be auto-detected).

>    of the target cluster.

Reviewer 3: We will label the volume in the ISO. A udev rule will detect this volume label and automatically trigger a systemd service parameterised with the volume mount point. The user doesn't need to know anything about anything. We have used this approach successfully in the past.

>    of the target cluster.

Reviewer 8: Is this method documented in this EP or any other EP? Might be worth a short note here to explain it

>    of the target cluster.

Author: I don't think there's any EP related to that. I will add more details below in the implementation details [Upgrade](https://github.com/openshift/enhancements/pull/1821/files#diff-7b20b50dcbb0b982828fe87930f63c142457b54311f85d0a73468c86ee2165c1R344) paragraph, where it is mentioned

4. The user edits the IRI resource by adding the release image pullspec to the
   `spec.releases` field.
5. The user waits until the copy process of the new release image is completed
   for all the control plane nodes. The progress and status of the task could
   be monitored via the IRI `status` field.
6. Once completed, the user upgrades the cluster via the `oc adm upgrade
   --allow-explicit-upgrade --to-image=<release-image>`

#### Post-upgrade optional steps

1. The user edits the IRI resource by removing the release image pullspec not

>    in use anymore.

Reviewer 3: Currently ~40GiB per release with all the OLM operators, but we have more OLM operators to add, so probably closer to 60. For just OCP, about 20.

>    in use anymore.

Reviewer 8: We need to be able to store at least 2 full payload copies right? To be able to handle current + upgrade?

In which case documenting that at least 120 GiB of storage is needed on each node seems quite important. I believe across most platforms we recommend 128GiB or less for the default volume size so this extra space is something to consider

>    in use anymore.

Reviewer 7: Indeed. It is important that we properly explain that the ~120GB (or whatever) is required for the releases and in addition to whatever disk space may be also required by workloads, etc.

Otherwise, some users may just install hard drives of 130GB, make them full with the 2 releases and complain when they cannot start 50 containers that do heavy usage of `EmptyDir volumes

>    in use anymore.

Author: Ok, I'll add an explicit node for the documentation (and for the live Assisted Service pre-flight check) related to the increased storage size required

>    in use anymore.

Reviewer 3: I don't disagree with any of this, but note that we're only storing the images on the control plane nodes.

>    in use anymore.

Reviewer 8: I don't see anything in the body of the EP currently documenting the additional disk overhead for control plane nodes, lets get something added and we can resolve this thread

>    in use anymore.

Author: Added

   configuration).
2. The user attaches the newly generate node ISO to the target machine(s),
   and boots it. The progress and status of the task can be monitored via the
   `oc adm node-image monitor` command.
3. The user waits until the CSRs are generated, and manually approves them
   via the `oc adm certificate approve` command.
   
### API Extensions

#### InternalReleaseImage custom resource

The `InternalReleaseImage` is a new singleton custom resource, controlled by
MCO, used to keep track and manage the OCP release images stored into the
control planes nodes. In particular, it will be used for:

* Triggering a new release image copy task - as a preliminary upgrade step
* Delete from disk a release image version not anymore in use - as an optional
  post-upgrade step
* Enabled/disable the registry services

This is an example of how the CR will look like:

```
apiVersion: releases.openshift.io/v1alpha1
kind: InternalReleaseImage
metadata:
  name: cluster
spec:
  releases:
    - "quay.io/openshift-release-dev/ocp-release:4.18.0-x86_64"
    - "quay.io/openshift-release-dev/ocp-release:4.19.0-x86_64"
  enabled: true

> status:

Reviewer 2: Would per-node status reporting be beneficial here? I would worry about multiple MCDs acting on a singular resource and trying to update the same status field causing conflicts.

Reviewer 3: If the CR is the source of truth, then per-node status reporting is essential.
However, if the controller can always ask MCD what the current status is then it could roll those up to a single status in the CR.
The enhancement for PinnedImageSet says that per-node reporting via the `MachineConfigNode` CRs will be "considered". I don't know what was actually implemented in the end.

Reviewer 4: It exists as a subfield on the MCN status: https://github.com/openshift/api/blob/master/machineconfiguration/v1/types_machineconfignode.go#L170

Reviewer 3: Thanks! Probably makes sense to do the same then I would think.

Author: Will prepare a first proposal

  releases:

>     - "quay.io/openshift-release-dev/ocp-release:4.18.0-x86_64"

Reviewer 4: Thinking about what we should report here, I wonder if each individual status entry for each payload should have subfields around progressing/success/error/etc, instead of overall conditions, so it's easier to see which payloads are there, which are being pulled, etc.

Although since a user is unlikely to specify more than 2 payloads here at a time, perhaps we just report the most recent error? Would the MCD would be constantly reconciling all releases it expects to be on disk? Or just the latest added/removed one?

Reviewer 4: In that case I think the status should have sub sections per release on the status. i.e. it would be nice to see how many nodes have successfully pulled each release, what errors there are, etc.

    - "quay.io/openshift-release-dev/ocp-release:4.19.0-x86_64"
  enabled: true
  conditions:
    - type: "Available"
      status: "True"
      lastTransitionTime: "2025-07-12T14:02:00Z"
      reason: "ReleasesSynchronized"
      message: "The previously managed releases are available."
    - type: "Progressing"
      status: "True"
      lastTransitionTime: "2025-07-12T13:13:00Z"
      reason: "AddingRelease"

>       message: "Adding new release image quay.io/openshift-release-dev/ocp-release:4.19.0-x86_64 on master-0."

Reviewer 3: I think adding and removing a release at the same time is fine. The limitation will be that you can't remove a release that the CVO is currently using (i.e. is the current target or the one we are in the process of upgrading from).

>       message: "Adding new release image quay.io/openshift-release-dev/ocp-release:4.19.0-x86_64 on master-0."

Reviewer 3: That would be the easiest way.
An alternative would be to say that we will independently look up what the current release image is, and if it appears in the Status do not remove it from the registry even if it has disappeared from the Spec.

The former is more explicit and therefore easier to explain to users I think.

>       message: "Adding new release image quay.io/openshift-release-dev/ocp-release:4.19.0-x86_64 on master-0."

Reviewer 7: Yep, I agree. If we silently keep the current release in the status when removed from spec, it would become necessary to constantly watch if it is still the current release or not and eventually schedule the removal. More complex.

I agree that just rejecting the update to the spec is easier.

  ...
```

* The `spec.releases` will be used to configure the desired release contents.
  Adding (or removing) an entry from the list will trigger an update on all the
  control plane nodes.

> * The `spec.enabled` will be used to turn on / off the feature. In particular,

Reviewer 2: Would be good to have a small blurb about `status.enabled` here too.

Reviewer 8: In that case, I would assume that customers will test the change on a staging environment right? If the new registry doesn't work out, will the existing images cached in the CRI still work? I guess unless they have `imagePullPolicy: Always`?

What would you expect time to recovery to be for rebuilding the local registry once the IRI is re-instated?

Author: > Do we actually expect users to frequently disable and then re-enable this? In https://github.com/openshift/enhancements/pull/1821#discussion_r2278399494 the suggestion is that it will be disabled for good?

ATM I do not expect any frequent feature opt-in/out (if not by mistake, because the user inadvertently removed the IRI resource), so I agree to delete everything when the resource is removed.

Reviewer 3: Just like we won't let people delete the current release image (the one that CVO is targeting or upgrading from) from the image list, we also won't let people disable the whole feature (by deleting the IRI CR) under the same circumstances.
So in order to disable the feature, the user will have to use CVO to upgrade (or 'upgrade') to a release in the external registry _first_. After that we can safely disable the feature and delete any images stored on disk.
(We should document this explicitly.)

Reviewer 3: [Yes](https://github.com/openshift/enhancements/pull/1821#discussion_r2282081134), that's what we decided.

  when turned off, all the registry services running on the control plane nodes
  will be stopped.
* The `status.releases` field will report the currently managed releases
* The `status.conditions` field will be used to keep track of the activity
  performed on the various nodes.

### Topology Considerations

Will function identically in all topologies.

#### Hypershift / Hosted Control Planes

N/A

#### Standalone Clusters

N/A

#### Single-node Deployments or MicroShift

N/A

### Implementation Details/Notes/Constraints

#### Extended RHCOS ISO build

The release image and operators images will be stored inside the extended RHCOS
ISO in the storage format used by the [distribution/distribution](https://github.com/distribution/distribution)
registry’s filesystem storage plugin.

> #### Distribution/distribution registry OCP integration

Reviewer 4: The image registry here has some interesting overlap with the MCO's image mode efforts, namely:
1. the design seems to be a lightweight, cluster-only registry used for cluster operations that we manage (somewhat through the MCO)
2. this registry could potentially be used for os image build distribution for this cluster

cc @pablintino who has been investigating image registry use cases for image mode

The [distribution/distribution](https://github.com/distribution/distribution)
registry will be used to run the registry services on each control plane

> node. The binary will be built and included in the [openshift/image-registry](https://github.com/openshift/image-registry)

Reviewer 3: `openshift/image-registry` is a downstream fork, and is part of the release payload. So it will be updated along with the release payload.
There is some extra complexity because this image is actually hosting the registry, so the image also needs to be obtained from the release payload and put into a containers/storage store where it is directly accessible by podman.

> node. The binary will be built and included in the [openshift/image-registry](https://github.com/openshift/image-registry)

Reviewer 3: Actually I was a bit inaccurate there. It's not a fork but a wrapper that vendors in the upstream distribution/distribution.
The [plan](https://github.com/openshift/image-registry/pull/426) is to add another binary to the container that is just building distribution/distribution from the vendor directory without any OpenShift-specific bits added.
That way any CVE fixes we apply automatically get included for both this purpose and the image-registry-operator purposes.

> node. The binary will be built and included in the [openshift/image-registry](https://github.com/openshift/image-registry)

Reviewer 3: > I guess the assumption here is that you'd also have a k8s service to dispatch to those?

I believe what we're saying is that we'll expose it on a host networking port, and it will be accessible via the api-int load balancer (in the case of platform:none we'll require users to add that port to their load balancer).

> repo image, so that it will be available within an OCP release image payload.

Reviewer 5: Which image in OCP payload will include the registry?
Will it be addressed as part of this effort, or do we have another enhancement for this?

#### Installation - bootstrap phase

![alt bootstrap phase](simplified-operations-bootstrap.png)

> Initially the registry image will be placed into a read-only [additional image

Reviewer 5: Currently we simply use 'podman load' on the registry tar file.
Are you referring to the same flow?

store](https://www.redhat.com/en/blog/image-stores-podman), configured to be
used by podman in `/etc/containers/storage.conf`.
An initial service (added by the Appliance builder in the ISO ignition) will
startup the registry at the very beginning of the boot phase. 
The `/etc/containers/registries.conf` will be modified to ensure that the
images will be pulled from the local registry, using `localhost` as the
registry host.

#### Installation - install phase

![alt install phase](simplified-operations-install.png)

> A systemd service injected by MCO will ensure the registry will be available

Reviewer 5: Same as the current approach in the appliance ([start-local-registry.service](https://github.com/openshift/appliance/blob/master/data/services/common/start-local-registry.service#L1))?

after the reboot. 
The registry will have to run on a host network port on each node to ensure
it's accessible from the outside, specifically port 22625 (to be [registered](https://github.com/openshift/enhancements/blob/master/dev-guide/host-port-registry.md)). 
An Assisted Service pre-flight validation will ensure that the port is open.

#### Upgrade

> A udev rule, along with a systemd service injected by MCO will detect when an

Reviewer 5: Something like the current [udev](https://github.com/openshift/appliance/blob/master/data/udev/rules.d/99-upgrade-iso.rules#L1) rule in the appliance, right?

> A udev rule, along with a systemd service injected by MCO will detect when an

Reviewer 8: Possibly both, but since the MCD is going to be looking at this (and is the closest openshift code to the mount), would it not make sense for it to be reported in the MCN (effectively each MCD gets an MCN to write information it knows about to)

extended RHCOS ISO is attached, an then mount it to launch a temporary registry
to serve its content.
After the user modified the IRI resource with the new release pullspec, the new

> OCP release image payload will be copied on each node registry using skopeo.

Reviewer 2: Perhaps a silly question, but is there value in these images being verified for integrity? For example, say the node goes through a reboot and the new MCD comes up and examines the IRI resource. How does it know the releases mentioned by the IRI spec field matches what is in its local registry?

Reviewer 3: Not a silly question!
Ideally, all images are referenced by digest rather than tag. That's enough to maintain integrity.
However, there's an obvious usability gap there, and how we bridge it is underspecified by this proposal.

Reviewer 3: I think we need to document each step from the perspective of the user:

* An ISO is attached, and a registry is made available
* The user has to explicitly acknowledge that they intended to do that by putting some data into the IRI resource. What data is that?
  * A version number? How will the system verify that the image it is pulling is the one that the user is explicitly acknowledging?
  * A pullspec with a tag? How will the system verify that the image it is pulling is the one that the user is explicitly acknowledging?
  * A pullspec with a digest? Where will the user copy the pullspec from? (And how will they tell which one to remove later when they are done with it?)
* The user then initiates an update with CVO. What pullspec will they use as the target? Where will the user copy the pullspec from? Is there any way to verify the [update graph](https://github.com/openshift/enhancements/pull/1821#discussion_r2278096071)?
  * The original quay pullspec with a tag? (I don't think this works because IDMS won't handle it and we don't want to use ITMS.)
  * The original quay pullspec with a digest?
  * A pullspec pointing directly to the mirror with a tag?
  * A pullspec pointing directly to the mirror with a digest?

> OCP release image payload will be copied on each node registry using skopeo.

Reviewer 6: Skopeo is pointed directly at the host service or does it also get an IMDS entry and skopeo thinks it's talking to quay.io? If it had an IMDS entry could the IRI controller be used, unaltered, in a connected environment because it would eventually synchronize from the origin?

Reviewer 3: There's something to be said for that, and also something to be said for not doing that because users who are connected but bandwidth-constrained will hate us if we download data over the network when they already supplied it locally :smile:

Either way, the answer needs to be made explicit in this proposal.

Reviewer 6: They'd have supplied it locally via an IMDS pointed at their local registry, right? I assume there's a way to say never even attempt to reach the origin and fail instead?

### Risks and Mitigations

* Exposing the internal registry on a node port will allow it to be reached
  from outside the cluster. This behavior is similar to the one already
  adopted by the Machine Config Server (MCS).
  This is a new requirement that will need to be properly documented. In
  addition, to minimize the risks, the registry will have to be secured

>   via a pull-secret.

Reviewer 3: We'll probably need to document how the user can force the credentials to be rotated.
I imagine that at installation time we'll generate credentials and save them in a Secret, as well as add the auth data to the cluster's pull secret.
If the user wants to rotate they'll need to delete the Secret (MCO will automatically replace?) and manually update the cluster pull secret.

>   via a pull-secret.

Reviewer 8: This sounds like an important workflow to document in the EP

>   via a pull-secret.

Author: Added some additional details for this part.

## Alternatives (Not Implemented)

* For building the ISO, we've explored also the usage of the [custom-coreos-disk-images](https://github.com/coreos/custom-coreos-disk-images)
  utility, to reuse the OCP layering feature. The approach was discarded as it
  appeared not feasible especially when dealing with such additional payload
  size (a standalone OCP release size is usually ~20GB)

* Instead of using a local registry on each node, we've also explored the
  possibility to pre-load all the required images into an additional read-only
  local container storage on each node. Unfortunately this approach wasn't
  feasible, as several pieces of the OpenShift ecosystem assume that a registry
  was always available in the environment. Removing such dependency across
  all the possible locations appeared to be particularly complex and
  challenging (and moreover no formal way to enforce it once removed)

## Test Plan

[dev-scripts](https://github.com/openshift-metal3/dev-scripts) will be enhanced
to support testing via the extended RHCOS ISO, and new jobs will be added in the
OpenShift CI to cover the sno/compact/ha topologies.

## Graduation Criteria

### Dev Preview -> Tech Preview

> - Ability to utilize the enhancement end to end

Reviewer 5: Hmm, do we know how to test it e2e?
E.g. use a config-image instead of the interactive 'below-sea-level' part?

      lastTransitionTime: "2025-07-12T13:13:00Z"
      reason: "AddingRelease"
      message: "Adding new release image quay.io/openshift-release-dev/ocp-release:4.19.0-x86_64 on master-0."
  ...
```

* The `spec.releases` contains a list of release objects, and it will be used
  to configure the desired release contents. Adding (or removing) an entry from
  the list will trigger an update on all the control plane nodes.
* The `status.releases` field will report the currently managed releases
* The `status.conditions` field will be used to keep track of the activity
  performed on the various nodes

### Topology Considerations

Will function identically in all topologies.

#### Hypershift / Hosted Control Planes

N/A

#### Standalone Clusters

N/A

#### Single-node Deployments or MicroShift

N/A

### Implementation Details/Notes/Constraints

#### InternalReleaseImage custom resource

A new `InternalReleaseImageController` sub-controller of the
_machine-config-controller_ will be created to watch and manage the
`InternalImageRelease` resource. The controller will take care of
updating the resource status.

A new `InternalReleaseImageManager` manager will be added to

> _machine-config-daemon_ to handle per-node specific operations.

Reviewer 8: What exactly would the spec be? What information are you expecting to pass to the MCD? (New annotation based APIs are generally forbidden, especially when we control the types)

> _machine-config-daemon_ to handle per-node specific operations.

Author: ATM (just) the list of the releases (to be copied from the ISO). In this way when a user edit the IRI resource to add/remove an entry in the release list, the IRI controller could start orchestrating the download/copy process on each node by updating the related MCN resource

   and boots it. The progress and status of the task can be monitored via the
   `oc adm node-image monitor` command.
3. The user waits until the CSRs are generated, and manually approves them
   via the `oc adm certificate approve` command.
   
### API Extensions

#### InternalReleaseImage custom resource

The `InternalReleaseImage` is a new singleton custom resource, named `cluster`,
controlled by MCO, used to keep track and manage the OCP release images stored
into the control planes nodes. In particular, it will be used for:

* Triggering a new release image copy task - as a preliminary upgrade step
* Delete from disk a release image version not anymore in use - as an optional
  post-upgrade step

This is an example of how the CR will look like:

```
apiVersion: releases.openshift.io/v1alpha1
kind: InternalReleaseImage
metadata:
  name: cluster
spec:
  releases:
    - name: quay.io/openshift-release-dev/ocp-release:4.18.0-x86_64
    - name: quay.io/openshift-release-dev/ocp-release:4.19.0-x86_64
status:
  releases:
    - "quay.io/openshift-release-dev/ocp-release:4.18.0-x86_64"
    - "quay.io/openshift-release-dev/ocp-release:4.19.0-x86_64"

>   currentRelease:

Reviewer 8: > So the list must be able to accommodate at least 3 items.

In the future, EUS upgrades may span larger drifts than even that, we can limit this to 3 now and expand later, but we should consider that this may need to be larger and the implications of that

Worth pulling in OTA for their thoughts on how many they might expect to actually be in use on one cluster at a time?

    - "quay.io/openshift-release-dev/ocp-release:4.18.0-x86_64"
  conditions:
    - type: "Available"
      status: "True"
      lastTransitionTime: "2025-07-12T14:02:00Z"
      reason: "ReleasesSynchronized"
      message: "The previously managed releases are available."
    - type: "Progressing"
      status: "True"
      lastTransitionTime: "2025-07-12T13:13:00Z"
      reason: "AddingRelease"
      message: "Adding new release image quay.io/openshift-release-dev/ocp-release:4.19.0-x86_64 on master-0."
  ...
```

* The `spec.releases` contains a list of release objects, and it will be used
  to configure the desired release contents. Adding (or removing) an entry from

>   the list will trigger an update on all the control plane nodes.

Reviewer 8: That would be my assumption yes, lets add a note that when this happens it will be reported (and a suggestion of what that status might look like would be good) please

`spec.releases` field of the `InternalReleaseImage` resource to delete the
desired release entry value.

A ValidatingAdmissionPolicy will be added to prevent the user to remove
a release entry that is currently being used by the cluster (as reported by
the `status.currentRelease` field).

##### Removal of InternalReleaseImage resource

The deletion of the IRI `cluster` resource will be handled by a finalizer,
and it will stop all the registry services running on the control plane nodes.
In addition, all the currently stored release payloads will be deleted from the
nodes disk, and all the previously resources created will be removed as well
(such as the IDMS ones) to complete the cleanup.

It will not possible to delete the resource if the current release is managed
by the IRI itself. To opt-out from the feature, the user will have to setup
his/her own registry with the required mirrored content and perform a cluster
upgrade using it.
Once successfully completed, the `InternalReleaseImageController` will 
blank out the `spec.currentRelease` field to indicate that none of the its
managed release payloads are currently being used for the cluster. At this

> point it would be possible for the user to remove the IRI resource.

Author: As per the previous [comment](https://github.com/openshift/enhancements/pull/1821#discussion_r2314237718) if VAP doesn't support cross-reference check this would be limited to the controller level

> point it would be possible for the user to remove the IRI resource.

Reviewer 8: See my response there, you can absolutely achieve this using a VAP

---
title: simplified-registry-less-cluster-operations
authors:
  - "@andfasano"
reviewers:
  - "@danielerez"
  - "@oourfali"
  - "@bfournie"
  - "@rwsu"
  - "@pawanpinjarkar"
  - "@yuqi-zhang"
  - "@mrunalp"
  - "@sdodson"
approvers:
  - "@zaneb"
api-approvers:
  - "@joelspeed"
creation-date: 2025-07-07
last-updated: 2025-07-07
tracking-link: 
  - "https://issues.redhat.com/browse/AGENT-1250"
  - "https://issues.redhat.com/browse/VIRTSTRAT-60"
see-also: []
replaces: []
superseded-by: []
---

# Simplified registry-less cluster operations for disconnected environments

## Summary

This enhancement proposal outlines an opinionated workflow to simplify
on-premises cluster operations - such as installation, upgrade and node
expansion - by eliminating the need to setup and configure an external
registry mirror. This approach is specifically designed for disconnected
environments and leverages existing tools such as Agent-based Installer,

> Assisted Installer and OpenShift Appliance.

Reviewer 9: I think I understand that this approach is simpler because it doesn't involve users setting up mirror registries on their disconnected network, so are we planning to funnel _all_ disconnected users to this approach?

If not, why would a customer need to stick with the traditional oc-mirror approach?
Or if so, should oc-mirror be able to directly create the ISO (and associated cluster objects) to avoid fragmentation of tooling, docs, etc.?

  disconnected environment without setting up an external registry (and
  optionally a selected set of OLM operators)
* As an OpenShift administrator, I want to upgrade a cluster into a
  disconnected environment without setting up an external registry (and
  optionally a selected set of OLM operators)
* As an OpenShift administrator, I want to add a node to an existing cluster in
  a disconnected environment without setting up an external registry
* As an OpenShift administrator, after having deployed a cluster, I want to
  start using my own external registries for future upgrades

### Goals

* Ensure successful installation/upgrade/adding new nodes in disconnected
  environments with minimal requirements and no pre-existing registry
* Offer a simplified installation experience for disconnected environments by
  leveraging existing tools like the Agent-Based Installer, Assisted Installer
  and OpenShift Appliance
* Support a mechanism to allow the user, after having successfully installed
  the cluster, to opt-out from the feature and start using his/her own external
  registry for the cluster upgrade operations

### Non-Goals

> * Allow the user to install/upgrade any OLM operators

Reviewer 9: I'm a little confused. I see user stories that talk about optionally installing/upgrading OLM operators, but then this is listed as a non-goal?

Reviewer 9: Ah, maybe `s/any/arbitrary/` ?

scripts/services to support the various operations. In general a local registry
service will be running on every control plane node, serving the OCP release
payload included in the ISO. Also, a new custom resource - managed by the 
Machine Config Operator (MCO) - will be introduced to control and manage the
internal release payload added to the system.

Since each cluster operation has its own requirements and characteristics, the
following paragraphs will provide more specific details for each different
scenario.

### Extended RHCOS ISO build and release

The foundation of the current proposal lies in the user's possibility to
download a customized self-sufficient RHCOS live ISO, containing all the
necessary elements to support the cluster operations in a disconnected
environment.
This ISO will be prepared and built using the [OpenShift Appliance Builder](https://github.com/openshift/appliance),
a command line utility for building a disk image to orchestrate an OpenShift
installation. The Appliance builder will be enhanced with a new 
`build live-iso` command capable to generate an ISO artifact. 
The extend ISO builder will be integrated within the official Red Hat build and

> release pipeline, so that a new extended RHCOS ISO could be published following

Reviewer 9: Will this official ISO include OLM operators? If so, which ones? And how do we choose? And of the ones we chose, are we mirroring all channels, bundles, and related images for those operators? For an official ISO, I think we would need to include the all channels, bundles, and related images to ensure that all possible upgrade paths remain intact. But that will bloat the overall ISO size.

Also, OLM catalogs have been intentionally decoupled from the release cadence of the OCP payload (customers can receive updates to their operators without upgrading their cluster). An official image that contains both an OCP payload and a snapshot of OLM operators implies that users of this method need to upgrade their cluster in order to upgrade their OLM operators. Should we have separate ISOs to keep these decoupled?
- One for payload (generated when new payload is released)
- One for OLM catalogs (generated when new catalogs are released, or on some periodic basis)

Author: Yes, the ISO will also include a subset of OLM operators. Including all the operators wouldn't be feasible due the huge size of the final artifact.  Even though the scope of the current proposal does not focus on a specific product - but more on the generic mechanism, it is related to https://issues.redhat.com/browse/VIRTSTRAT-60 (as reported initially) where you could find a more complete list of the included operators.

that it remains available not only after the first reboot, but also once the
installation is complete. Also the registry container image itself will be
copied in the node additional image store for podman.
[Assisted Service](https://github.com/openshift/assisted-service/)
and [Assisted Installer](https://github.com/openshift/assisted-installer)
will be enhanced accordingly to take care of this step, and to ensure it
will be properly monitored during the installation process.

A new systemd service will be injected via MCO to run the registry container
for providing the release content stored on disk after the reboot (since the
previous ephemeral environment will not be available anymore).

To allow pulling images from the cluster, an ImageDigestMirrorSet
(IDMS) policy will be added to the cluster manifests - using the `api-int` 
DNS entry - so that the set of registries running within
the control plane nodes will transparently serve the release payload images
locally mirrored.

Finally, a new `InternalReleaseImage` custom resource, managed by MCO, will
also be added as a cluster manifest to keep track and manage the currently
stored internal release image.

> 

Reviewer 9: Add details of the OLM aspect of the installation here?

Author: Do you maybe mean reporting the list of what was exactly included in the current release bundle? If so I will sooner update the proposal with more details on this topic (see also this [comment](https://github.com/openshift/enhancements/pull/1821/files#r2342887223))

Author: cc @danmanor for additional comments

### Upgrade

Upgrading an existing lower version cluster will be performed via the same
extended ISO used for managing the installation. As soon as the user will
attach the extended ISO to one of the control planes a temporary registry
will be created to serve its content.
At this point the user could edit the existing IRI resource to add the new
release to trigger the copy process (from the temporary registry
to each local control plane node registry), managed by the Machine Config
Daemon (MCD). Once the copy has been completed on all the control plane nodes,
the user could start the usual offline upgrade process via the `oc adm upgrade
--allow-explicit-upgrade --to-image=<release-image>`.
At the end of the upgrade the user may decide to delete the previous release
entry from the IRI resource. In such case the MCD on each control plane node
will take care of removing the older release payload from the local registry.

> 

Author: This is definitely a good point and something we'd like to elaborate further, since AFAIU the upgrade path for an OLM operator may be more complicated - so any feedback / suggestion is welcomed.

ISO (bootstrap phase).
In addition to the usual ABI services configured to orchestrate the
installation, a local registry service will be created on each node, and the
content of the release payload from the mounted ISO will be used as its
backend source.
During this early phase, the main responsibility of the local registry will
be to serve any image pull request originating from the local services running
on the same node.

##### Installation phase (after the first reboot)

During the installation of each control plane node, the contents of the OCP
release payload will be copied from the ISO to the node's disk (in the format
of the registry's 'filesystem' backend) after the OS disk writing step - so
that it remains available not only after the first reboot, but also once the
installation is complete. Also the registry container image itself will be
copied in the node additional image store for podman.
[Assisted Service](https://github.com/openshift/assisted-service/)
and [Assisted Installer](https://github.com/openshift/assisted-installer)
will be enhanced accordingly to take care of this step, and to ensure it

> will be properly monitored during the installation process.

Author: IMO it's not a nice-to-have feature but a critical step to make the installation workflow more robust and properly monitored. 
Copying 40Gb+ of data takes a good amount of time on a baremetal instance (up to 1hr as per the current testing), and several things can go wrong during the process, with the potential risk of leaving the data uncompleted or corrupted (as already experienced in the 4.19 DP release) - and thus the user should get properly informed that everything went fine (or if some problem happened, to avoid having later issues during the joining process, usually very painful to troubleshoot).
During the installation phase the Assisted Installer is the current eligible actor, on each node, responsible to carry on all the required task (such as writing the RHCOS image), and monitoring it's already supported via communication with the Assisted Service, which could be used by the user (via openshift-install wait-for command, or via the local Assisted UI) for that. And in this way we could also ensure that, prior of rebooting a node to let it join the control plane, all the required data (for making it working) will be effectively available.

> A new systemd service will be injected via MCO to run the registry container

Reviewer 10: An operator leveraging the static pod controller would be preferable here. This would be more in line with what the etcd and kube-apiserver operators do.

  restricted to a specific subset.

## Proposal

From a broad perspective, this proposal consists of building and releasing an
extended RHCOS live ISO that includes a full OCP release payload (with a
selected subset of OLM operators) and all the necessary scripts/services to
support the various operations. In general a local registry service will be
running on every control plane node, serving the OCP release payload included
in the ISO. Also, a new custom resource - managed by the Machine Config
Operator (MCO) - will be introduced to control and manage the internal release
payload added to the system.

### High-level proposal scenarios

#### Extended RHCOS ISO build and release

The foundation of the current proposal lies in the user's possibility to
download a customized self-sufficient RHCOS live ISO, containing all the
necessary elements to support the cluster operations in a disconnected
environment.
This ISO will be prepared and built using the [OpenShift Appliance Builder](https://github.com/openshift/appliance),
a command line utility for building a disk image to orchestrate an OpenShift
installation. The Appliance builder will be enhanced with a new 

> `build live-iso` command capable to generate an ISO artifact.

Reviewer 11: ```suggestion
`build live-iso` command capable for generating an ISO artifact.
```

Since the extended ISO will contain not only a specific set of OCP release images

> - but also additional images related to OLM operators (and other support images),

Reviewer 11: Nit - the '-' at the beginning of the line causes a bullet to be generated. Should change to a comma.

   of the target cluster.
4. The IRI resource will report the newly detected release bundle
   identifier with a `Mounted` condition in its `status.releases` field.
5. The user edits the IRI resource by adding the release bundle identifier to
   the `spec.releases` field.
6. The user waits until the copy process of the new release image is completed
   for all the control plane nodes. The progress and status of the task could
   be monitored via the IRI `status` field.
7. Once completed, the IRI status will report also the specific release
   pullspec, so that the user could upgrade the cluster via the `oc adm upgrade
   --allow-explicit-upgrade --to-image=<release-pullspec>`

#### Post-upgrade optional steps

1. The user edits the IRI resource by removing the release bundle identifier
   not in use anymore.
2. All the images related to the specified release bundle will be removed
   from all the control plane nodes. The progress and status of the task
   could be monitored via the IRI `status` field.

#### Node expansion

1. The user runs the `oc adm node-image create` command from within the
   disconnected environment (with optionally some additional node
   configuration).

> 2. The user attaches the newly generate node ISO to the target machine(s),

Reviewer 11: ```suggestion
2. The user attaches the newly generated node ISO to the target machine(s),
```

   and boots it. The progress and status of the task can be monitored via the
   `oc adm node-image monitor` command.
3. The user waits until the CSRs are generated, and manually approves them
   via the `oc adm certificate approve` command.
   
### API Extensions

#### InternalReleaseImage custom resource

The `InternalReleaseImage` is a new singleton custom resource, named `cluster`,
controlled by MCO, used to keep track and manage the OCP release images stored
into the control planes nodes. In particular, it will be used for:

* Triggering a new release image copy task - as a preliminary upgrade step

> * Delete from disk a release image version not anymore in use - as an optional

Reviewer 11: ```suggestion
* Delete from disk a release image version no longer in use - as an optional
```

  post-upgrade step
* Opt-out from the feature (by deleting the resource)

> This is an example of how the CR will look like:

Reviewer 11: ```suggestion
This is an example of what the CR will look like:
```

```
apiVersion: machineconfiguration.openshift.io/v1alpha1
kind: InternalReleaseImage
metadata:
  name: cluster      
spec: 
  releases:
    - name: ocp-release-bundle-4.18.0-x86_64
    - name: ocp-release-bundle-4.19.0-x86_64
status:
  conditions:
    - type: Degraded
      status: "False"
      reason: "AllReleasesAvailable"
      message: "All the release images are available"
      lastTransitionTime: "2024-11-01T07:00:00Z"
  releases:
    - name: ocp-release-bundle-4.18.0-x86_64

>       image: example.com/example/openshift-release-dev@sha256:aa8795f7932441b30bb8bcfbbf05912875383fad1f2b3be08a22ec148d6860ff

Reviewer 11: Should use real value of `api-int` here instead of `example.com` to make it clearer.

      conditions:
      - type: Available
        status: "True"
        reason: "Available"
        message: "Release ocp-release-bundle-4.18.0-x86_64 is currently available"
        lastTransitionTime: "2024-11-01T07:00:00Z"
    - name: ocp-release-bundle-4.19.0-x86_64

>       image: example.com/example/openshift-release-dev@sha256:d98795f7932441b30bb8bcfbbf05912875383fad1f2b3be08a22ec148d68607f

Reviewer 11: Same comment about using `api-int`

      conditions:
      - type: Available
        status: "True" 
        reason: "Available"
        message: "Release ocp-release-bundle-4.19.0-x86_64 is currently available"
        lastTransitionTime: "2024-12-01T09:00:00Z"
```

* The `spec.releases` contains a list of release objects, and it will be used
  to configure the desired release contents. Adding (or removing) an entry from
  the list will trigger an update on all the control plane nodes
* The `spec.releases.name` format is a string specifying the release bundle
  identifier. It must match the value discovered and reported in the 
  `status.releases` field
* The `status.conditions` field will be used to report the current global state
* The `status.releases` field will report the list of each currently managed release
* The `status.releases.name` is the identifier of the release bundle
* The `status.releases.image` is the OCP release image pullspec by digest for
  the current release bundle

> * The `status.conditions` field will be used to keep track of the activity

Reviewer 11: Should this be` status.releases[].condition`?

that it will be possible to track the current status/errors of each individual
node. The IRI resource will monitor all the MCN resources to update its status.

Following a proposal example for the extended status field:

```
apiVersion: machineconfiguration.openshift.io/v1
kind: MachineConfigNode
...
status:
  internalReleaseImage:
    releases:
      - name: ocp-release-bundle-4.18.0-x86_64
        image: example.com/example/openshift-release-dev@sha256:aa8795f7932441b30bb8bcfbbf05912875383fad1f2b3be08a22ec148d6860ff
        conditions:
          - type: Mounted
            status: "False" 
            reason: "Mounted"
            message: ""  
            lastTransitionTime: "2024-12-01T08:04:21Z"
          - type: Available
            status: "True"               
            reason: "Available"
            message: "Release ocp-release-bundle-4.18.0-x86_64 is currently available on node master-0"
            lastTransitionTime: "2024-12-01T08:04:21Z"
          - type: Degraded
            status: "False"               
            reason: "Degraded"
            message: ""
            lastTransitionTime: "2024-12-01T08:04:21Z"
```

> * The `internalReleaseImage.status.releases` field will be used to track the status of the release

Reviewer 11: The format is actually `status.internalReleaseImage.XXX` not `internalReleaseImage.status.XXX`

Author: ah good catch

  payloads stored in the related node
* The `internalReleaseImage.status.releases.name` will report the release bundle identifier string
* The `internalReleaseImage.status.releases.image` will report the OCP release image pullspec by digest
* The `internalReleaseImage.status.conditions` field will be used to keep track of the activity
  performed for the specific release payload, and in particular to report any error

For each release bundle, the following bundle condition types will be defined:

* `Mounted`. When true, it means that a valid ISO has been mounted on the current node.
* `Installing`.  When true, it means that a new release bundle is currently being copied on the current node, and not yet completed.
* `Available`. When true, it means that the release has been previously installed on the current node, and it can be used.
* `Removing`.  When true, it means that a release deletion is in progress on the current node, and not yet completed.
* `Degraded`. When true, it means something has gone wrong in the current node.

> Additionaly, a new `StateProgress` enum value will be added to the MachineConfigNode conditions type:

Reviewer 11: ```suggestion
Additionally, a new `StateProgress` enum value will be added to the MachineConfigNode conditions type:
```

N/A

#### Standalone Clusters

N/A

#### Single-node Deployments or MicroShift

Will support SNO deployments.

### Implementation Details/Notes/Constraints

#### InternalReleaseImage custom resource

A new `InternalReleaseImageController` sub-controller of the
_machine-config-controller_ will be created to watch and manage the
`InternalReleaseImage` resource. The controller will take care of
updating the resource status.
In general, the IRI controller will watch all the available
MachineConfigNodes status to verify if a given action (adding/removing a
release bundle) has been fully completed, and to detect the presence
of a new mounted ISO.

A new `InternalReleaseImageManager` manager will be added to
_machine-config-daemon_ to handle per-node specific operations.
Every manager will report its own status to the related MachineConfigNode
resource.
Every manager will be responsible for watching the IRI resource, to detect
when adding/removing a release bundle is required.
Also, if a release bundle ISO is mounted on the related node, it will have
to report the ISO label string into its status.

##### Removal of a release from the InternalReleaseImage resource

> The user could remove a release payload not anymore in use by editing the

Reviewer 11: ```suggestion
The user could remove a release payload no longer in use by editing the
```

[registered](https://github.com/openshift/enhancements/blob/master/dev-guide/host-port-registry.md)). 
An Assisted Service pre-flight validation will ensure that the port is open.

#### Upgrade

A udev rule, along with a systemd service injected by MCO will detect when an
extended RHCOS ISO is attached, and then mount it to launch a temporary registry
to serve its content, by looking for devices labelled as `ocp-release-bundle-*`.
This label will be reported as the bundle identifier in the MCN status (and then
in the IRI status). After the user modified the IRI resource with the new release
bundle identifier, the new OCP release image payload will be copied on each node
registry using skopeo.

#### Storage requirements

The control plane nodes disks must have sufficient space to store the requested
release payloads. The recommendation is to allocate at least ~60Gb per release.
Thus, to fully support also the upgrade, a total of additional ~120GB will be
required.
Assisted Service will apply a pre-flight validation to ensure that the
control plane nodes will meet the new storage requirement. Additionally, it

> should be properly documented in the pre-requistes section.

Reviewer 11: ```suggestion
should be properly documented in the prerequistes section.
```

### Risks and Mitigations

* Exposing the internal registry on a node port will allow it to be reached
  from outside the cluster. This behavior is similar to the one already
  adopted by the Machine Config Server (MCS).
  This is a new requirement that will need to be properly documented. In
  addition, to minimize the risks, the registry will have to be secured
  via a pull-secret.

### Drawbacks

* The extended RHCOS ISO size is pretty big, currently ~50GB with just a
  limited set of OLM operators. Including all the OLM operators is not
  feasible, and the current selected set may not be sufficient and/or
  effectively required by all the users.

> * Due the size, some hardware vendors may have difficulties into loading

Reviewer 11: ```suggestion
* Due to the size, some hardware vendors may have difficulties in loading
```

  such ISO, so a more extended testing may be required.
* The lifecycle of some OLM operators is different from the OCP one,

>   thus assembling an extended ISO may be challenging due the different release

Reviewer 11: ```suggestion
  thus assembling an extended ISO may be challenging due to the different release
```

  local container storage on each node. Unfortunately this approach wasn't
  feasible, as several pieces of the OpenShift ecosystem assume that a registry
  was always available in the environment. Removing such dependency across
  all the possible locations appeared to be particularly complex and
  challenging (and moreover no formal way to enforce it once removed)

## Test Plan

[dev-scripts](https://github.com/openshift-metal3/dev-scripts) will be enhanced
to support testing via the extended RHCOS ISO, and new jobs will be added in the
OpenShift CI to cover the sno/compact/ha topologies.

## Graduation Criteria

### Dev Preview -> Tech Preview

- Ability to utilize the enhancement end to end
- End user documentation
- Sufficient test coverage (using a disconnected environment)
- Gather feedback from users rather than just developers
- Integration with the official Red Hat build and release pipeline

### Tech Preview -> GA

- Complete e2e testing on installation, upgrade and node expansion
- User facing documentation created in [openshift-docs](https://github.com/openshift/openshift-docs/)

### Removing a deprecated feature

N/A

## Upgrade / Downgrade Strategy

> N/A

Reviewer 11: Seems that there should be some description in this section since the doc covers upgrade/downgrades, or maybe just refer to the relevant section.
